### PR TITLE
Expansion/legion

### PR DESCRIPTION
**Pull request by vendethiel**
First commit contains my changes
Second commit contains the libs.

I really want to ditch .pkgmeta for many reasons, most of which I explained in #1 already.
.pkgmeta doesn't help with basically anything – it doesn't provide versioning, it has stale references to everything, and it encourages using a tool that no one uses. If anything I'd rather use git submodules.
I also had to patch LibCooldownTracker to add per-spec cooldown overrides like legion has.

### CODE REVIEW

**Comment by vendethiel** [comment]
Looks like the whitespace changes (huh?) are really messing this up.


[Better viewed with Github's `?w=1` option.](https://github.com/slaren/GladiusEx/pull/2/commits/ae3f6daa6898635999cdce0e80d5273f3bb5dfb2?w=1)

**Comment by slaren** [comment]
Different type of line endings maybe ?

**Comment by vendethiel** [comment]
I think so too.

**Comment by vendethiel** [comment]
I only linked to the first commit btw, the second one has the libs and is not too interesting. 

[The hack in LibCooldownTracker is here.](https://github.com/vendethiel/GladiusEx/blob/112416d962c3b2371571df73b0932c49f1966c4c/libs/LibCooldownTracker-1.0/library.lua#L171-L182)

**Comment by vendethiel** [comment]
Uhm, btw... It seems WowAce's registration module is disabled. No clue what to do there..

**Comment by slaren** [comment]
Can you try to fix the whitespaces issue ? You obviously spent a lot of time on this and I kinda feel bad for asking this but I would rather not break the change history. Github also won't let me add a comment for a specific line in the ?w=1 pages.

The .pkgmeta stuff is good for keeping libraries automatically updated. It was present in wowace at a time where git wasn't that popular and most addons used subversion instead, but I guess nowadays git submodules may be a better option. Really, it only makes a different for the maintainer (less effort to keep the libraries updated), and since currently the addon is unmaintained, copying the libs in the repo is better than the alternative. However I would highly recommend using one of the two options (.pkgmeta or submodules) rather than just manually copying the libs to the repository, otherwise keeping them updated is going to be a major waste of time.

**Comment by vendethiel** [comment]
> Can you try to fix the whitespaces issue ? You obviously spent a lot of time on this and I kinda feel bad for asking this but I would rather not break the change history. Github also won't let me add a comment for a specific line in the ?w=1 pages.

I feel the same, yeah.

**Comment by vendethiel** [comment]
Ugh, this is gonna be super annoying.
I probably will remove my commits and force push a new version -_-.

**Comment by vendethiel** [comment]
Much better! Now it's all good.

I mean, I lost my own history, but I did things the wrong way, so no big deal.

File: GladiusEx.lua [hunk 1]
```diff
@@ -636,23 +636,29 @@ end
...
+	self:ShowFrames()
 
 	local numOpps = GetNumArenaOpponentSpecs()
 	for i = 1, numOpps do
 		local specID = GetArenaOpponentSpec(i)
 		local unitid = "arena" .. i
 
-		self:UpdateUnitSpecialization(unitid, specID)
+		if specID and specID > 0 then
+			self:ShowUnit(unitid)
```
Comment (vendethiel) [inline GladiusEx.lua]: That's probably a bit overkill...

File: GladiusEx.lua [hunk 2]
```diff
@@ -816,11 +822,14 @@ function GladiusEx:CheckUnitSpecialization(unit)
 end
 
 function GladiusEx:UpdateUnitSpecialization(unit, specID)
+	log("updateunit:"..unit)
 	local _, class, spec
```
Comment (vendethiel) [inline GladiusEx.lua]: I can probably remove `class` from here/cleanup the `GetSpecializationInfoByID` below.

File: GladiusEx.toc [hunk 3]
```diff
@@ -1,8 +1,12 @@
-﻿## Interface: 70000
+﻿## Interface: 70200
 ## Title: GladiusEx
 ## Notes: Arena unit frames
 ## SavedVariables: GladiusExDB
 ## OptionalDeps: LibStub, LibFunctional-1.0, Ace3, DRData-1.0, LibSharedMedia-3.0, AceGUI-3.0-SharedMediaWidgets, AceGUI-3.0-Spell-EditBox, LibGroupInSpecT-1.1, LibCooldownTracker-1.0, LibDualSpec-1.0, LibRangeCheck-2.0, LibDispellable-1.0, Clique, Masque
+## X-Curse-Packaged-Version: v0.35-beta-4-ga06b19c
```
Comment (vendethiel) [inline GladiusEx.toc]: I'll change this to release (why is the commit hash here?)

File: modules/alerts.lua [hunk 4]
```diff
@@ -24,7 +24,7 @@ local function GetDefaultCastsSpells()
 		[GladiusEx:SafeGetSpellName(105421)]   = cc,    -- Blinding Light
 		[GladiusEx:SafeGetSpellName(20066)]    = cc,    -- Repentance
 		[GladiusEx:SafeGetSpellName(33786)]    = cc,    -- Cyclone
-		[GladiusEx:SafeGetSpellName(102051)]   = cc,    -- Frostjaw
+		--V: removed in legion [GladiusEx:SafeGetSpellName(102051)]   = cc,    -- Frostjaw
```
Comment (vendethiel) [inline modules/alerts.lua]: Whoops. I need to cleanup.

File: modules/classicon.lua [hunk 5]
```diff
@@ -348,7 +396,7 @@ function ClassIcon:SetClassIcon(unit)
 			local n = 9 / 64
 			self.frame[unit].portrait2d:SetTexCoord(n, 1 - n, n, 1 - n)
 		end
-		if not UnitIsVisible(unit) or not UnitIsConnected(unit) then
+		if false then -- and not UnitIsVisible(unit) or not UnitIsConnected(unit) then
```
Comment (vendethiel) [inline modules/classicon.lua]: oops.

File: modules/unitbar.lua [hunk 6]
```diff
@@ -63,6 +63,10 @@ function GladiusEx:NewUnitBarModule(name, defaults_arena, defaults_party)
 			self.frame[unit].statusbar:SetStatusBarColor(color.r, color.g, color.b, color.a or 1)
 			self.frame[unit].icon:SetTexture([[Interface\Glues\CharacterCreate\UI-CharacterCreate-Classes]])
 
+			-- V: wtf is "CLASS_BUTTONS"?
+			if not CLASS_BUTTONS then
```
Comment (vendethiel) [inline modules/unitbar.lua]: more investigation needed, for sure... later.

**Comment by vendethiel** [comment]
Just as a reminder of sorts – this is the first time I do anything major on an addon, so... Probably looks a bit gory at times.

File: GladiusEx.toc (hunk 3 above)
Comment (slaren) [inline GladiusEx.toc]: That's added automatically by the wowace packager.

File: GladiusEx.lua (hunk 1 above)
Comment (slaren) [inline GladiusEx.lua]: To be honest the whole logic of when to show/hide units is a mess.

File: GladiusEx.toc (hunk 3 above)
Comment (vendethiel) [inline GladiusEx.toc]: Should i remove these curse lines?

File: modules/unitbar.lua (hunk 6 above)
Comment (slaren) [inline modules/unitbar.lua]: IIRC, **Interface\Glues\CharacterCreate\UI-CharacterCreate-Classes** was a texture used to contain all the class icons in a single texture. CLASS_BUTTONS used to be a global from the wow UI that had the texture coordinates for each class. I don't know if it has been renamed to something else or entirely removed.

File: GladiusEx.toc (hunk 3 above)
Comment (slaren) [inline GladiusEx.toc]: Yes, let the wowace packager add them, if you are planning on using it at all. It is probably used by the curse addon updater to keep track of what versions are installed.

Comment (vendethiel) [inline GladiusEx.toc]: I don't know enough about how wowace packages work yet. I'll need to fumble around a bit.

File: modules/unitbar.lua (hunk 6 above)
Comment (vendethiel) [inline modules/unitbar.lua]: Oh, okay. I thing this is just related to backgrounds being removed from such places (same for GetSpecializationInfoByID).
I can confirm the class icon correctly shows.

File: modules/unitbar.lua [hunk 7]
```diff
@@ -62,22 +62,6 @@ function GladiusEx:NewUnitBarModule(name, defaults_arena, defaults_party)
 
 			self.frame[unit].statusbar:SetStatusBarColor(color.r, color.g, color.b, color.a or 1)
 			self.frame[unit].icon:SetTexture([[Interface\Glues\CharacterCreate\UI-CharacterCreate-Classes]])
-
-			-- V: wtf is "CLASS_BUTTONS"?
-			if not CLASS_BUTTONS then
```
Comment (slaren) [inline modules/unitbar.lua]: This won't work if you just remove all of this. Looking at how it is done in classicon.lua, CLASS_BUTTONS was renamed to CLASS_ICON_TCOORDS.

Comment (vendethiel) [inline modules/unitbar.lua]: well, the code has been `return`ing early for all my testing, and I never had an issue with class icon not showing up.

Comment (vendethiel) [inline modules/unitbar.lua]: Ah, it's for the texture behind? Currently it only shows the color, it's true.

Comment (slaren) [inline modules/unitbar.lua]: unitbar.lua is kind of a template for the target bar and pet bar modules. It would break the class icon in these modules.

Comment (vendethiel) [inline modules/unitbar.lua]: Okay. Let's fix this.

**Comment by Rainrider** [comment]
AFAIK, you can't use gitmodules with svn repos. The ACE libraries are still on svn.

Apart from that, .pkgmeta would allow you to offer a disembedded installation option and the use of replacement tokens, which are nice for versioning. You could also generate your curse/wowace page from a file in the repo. See https://www.curseforge.com/docs/packaging

**Comment by vendethiel** [comment]
Generating a git mirror for a svn repo is very easy, however, github deals with that nicely.
I'll consider keeping the Ace libs tho. I won't need to change them in the foreseeable future.

What use is there offering a disembedded installation?

**Comment by Rainrider** [comment]
You still have to update the git mirror right? It can be automated but why the hassle, when you can have the whole thing done for you by a single simple file?

Disebedded installs are useful for developers interested in memory consumption. With embedded installs the memory foot print of the lib will be blamed to the addon that loaded it first. The other two benefits of disembedding are mostly ignored by users - only one instance of the lib is installed at once (so no wasted disk space and no multiple version comparisons/loads by LibStub) and they can upgrade libs independently without you having to push a new version of your addon. Also, the normal development workflow should be disembedded.

**Comment by vendethiel** [comment]
@Rainrider  You do have to update a git mirror, but adding a cron to do that would be very basic.
Maybe you know if I can pick a middle-of-the-road approach, and have some libs in the repo, as well as some via .pkgmeta?
I develop directly in my Interface folder (I don't mind since it's git'd anyway), so I'd have to ignore out some directories if I were to go for that – Disembedded looks like overkill for addon development. I see why it's there, but I'm not too worried about having Ace3 twice on disk.

@slaren any other notes? If I need to make other changes I'll only be able to test them in a week when I get back.

**Comment by Rainrider** [comment]
@vendethiel I'll just share my experience, you decide if it suits your workflow or not.

AdiButtonAuras (ABA) depends on several libraries, I'll use LibStub (LS), LibPlayerSpells-1.0 (LPS) and LibSpellbook-1.0 (LSB) for the example. My personal oUF layout also uses LS and LPS. All of them are separate addons in my Interface/AddOns folder = I run them disembedded. I maintain LPS and LSB, so I push changes to them and It makes a lot of sense to me to have only one instance of them on my drive instead of going the git submodules route. Github holds the dev repos for all of those (except for LS, which is on wowace and uses SVN). ABA and LPS are hosted (just the packages not the git repos) on wowace/curse and are auto-packaged by their system. LSB is only hosted on Github, but the curse packager can pull form every openly accessible svn/git/hg repo out there, it just needs the url and the relative path in your addon where it should extract the repo, and you list those in .pkgmeta.

I use the curse (now twitch) client for updating the addons I don't contribute to. LS is one of them. It detects addons I modified and does not auto-update them.

So, I can modify, maintain, distribute and update all of my addons in a consistent manner. All I have to do for this is edit the .toc and .pkgmeta files of my own addons, where it is two extra lines per toc file (only if you want to support disembedded installs for end-users, else you don't need to change anything there) and one line per addon dependency in the .pkgmeta (plus one extra line to disable disembedded installs if you want that).

In your case, you could just edit your current .pkgmeta (the repo urls it has are outdated) or convert from svn to git, write your cronjobs and use git submodules.

**Comment by vendethiel** [comment]
That's much cleaner now. thanks for the detailed explanation.

**Comment by slaren** [comment]
I agree, and seeing that git submodules aren't really an option I would be much more confortable with fixing the .pkgmeta instead of just manually inserting the libs here.

**Comment by vendethiel** [comment]
If I can keep LibDRData and LibCT in the repo while using pkgmeta for the rest, maybe.
I'll definitely sound a bit ranty but...It's still mighty annoying as a developer (because I'll need to gitignore my own versions or use the @no-lib-strip hack+disembedded; and keep them up-to-date for myself) and I very very much prefer to have pinned deps rather than "svn commits are our versions" which Ace3 seems to do but whatever.

fwiw, the base Gladius discarded .pkgmeta a good while ago, and just has the libs in their github repo. They only update libs when they do want a new feature or smth broke, rather than having to pin to a svn commit or using "latest" and praying they don't break backwards compat.

Anyway, the earliest testing I can do is next week.

**Comment by slaren** [comment]
The addon should be compatible with the latest version of Ace3 anyway (same for other libs too really), since any other addon could be embedding a more recent version and it would override ours. It's fine to manually embed libs like DRData if they are no longer maintained. However I am not sure why you want to disembed LibCT, just push your changed to its repository.

For what it's worth, the way I would develop GladiusEx back in the day is with the libs disembedded. I would just let the curse client/whatever keep the the libs updated to make sure that I would always be developing with the latest versions. That worked for me, and other than once in a while having to test the automatically packaged version to make sure that embeds were working fine it had very little overhead. Certainly much less than it would have been to have to manually update every lib.

Anyway, I know it is more work to fix it now and if you don't want to do it I won't press on it. Just let me know and I'll accept the PR and give you permissions to the wowace project so that you can upload a .zip or whatever.

**Comment by slaren** [comment]
About Gladius, I don't think they ever used a .pkgmeta. I don't think they even used Ace back when I forked it. A significant amount of time was spent adapting it to my standards.

**Comment by vendethiel** [comment]
I had to add a hack to LibCT to make it gladiusex-aware, because I wanted to be able to overload cooldowns based on class/spec (Legion talents that reduce CD etc often do not create a new spellID, annoyingly). That's the reason why I want to keep it here. 

If you say you were developing it disembedded, then I guess I can try. I just need to figure out how to package it locally so I can do "dry runs" and confirm the packaging works(like you said).
I don't really want to use curse (memory hog and I don't trust them too much) but I guess it's better than a perl script to update all the svn repos...

Thanks for spending all that time explaining stuff, both of you :-)

**Comment by slaren** [comment]
You don't need to package it yourself, just let the wowace packager do it (it should automatically generate a packaged .zip every time a commit to master is pushed) and download it. It's fine if the packaged .zip is broken for a few versions until you get it right, that's what alphas are for anyway. Once you get it right make a tag and a beta will be generated. However as far as I know the .pkgmeta system is exclusive to wowace/curseforge so I don't think you can avoid using curse.

**Comment by slaren** [comment]
I would also prefer to keep LibCT in its own repository even if it now depends on GladiusEx. You could keep the changes in a different fork but I don't think anyone is using it anyway.

**Comment by vendethiel** [comment]
True, I could keep it in a diff. repo. Would just need to make it very explicit that it's for gladiusex only.

**Comment by Rainrider** [comment]
I only had a quick look as you don't have a clean history and it is kind of PITA to track changes, but you added one single function to LCT that depends on GladiusEx - GetCooldownTime - and it returns data even when GladiusEx is not found. And you only have one case where this function will return different data with GladiusEx - subtlety rogues using Thief's Bargain, which you always assume anyways, so why bother with this and not just set the basic cooldown to 30 sec for sub rogues? So LCT is not dependent on GladiusEx either way.

**Comment by vendethiel** [comment]
LCT doesn't have spec info, AFAIK, that's why I need this.
(I also assume all discs run with fear CD talent)

**Comment by Rainrider** [comment]
I see https://github.com/slaren/LibCooldownTracker-10/blob/master/cooldowns_hunter.lua#L175-L180 in the old version, so I assumed it has it.

The only use of `cooldown_overload` in your fork is in the rogue data of LCT. Maybe you haven't uploaded it yet. But even then, it is the same use case as for sub rogues - you assume it is always used, because you can't know the talents of your pvp opponents.

Btw LCT data layout could be improved. Listing the class in every spell info table is a waste. Why not organize by class, spec, category? You could also move from associative to indexed arrays too (the categories might be a problem here though, but it can be solved by bitmasks), You could also have duplicates there, so if you load spec data after you have loaded the spells shared by all class specs, you could overwrite the shared spell with the spec specific one without introducing new fields like `cooldown_overload`.

**Comment by vendethiel** [comment]
The reason I have to use the _overload is because the arrays are indexed by spellid. Sub rogue's vanish and other spec's vanish share spellid, but not thief's bargain.

That doesn't solve the issue at hand tho. I still need to know, from libct, what's the player's spec.

I agree the layout could be improved, but I don't want to get into that yet, and not in this PR.

**Comment by vendethiel** [comment]
When I say "LCT doesn't have spec info", I mean it doesn't know the spec of whoever casted the spell.
That's the interesting part. The specID is just used for display purposes in gladiusex.

**Comment by slaren** [comment]
The spells are indexed by spellid mainly because it is the way they are looked up in the places where performance is a concern (in event handlers). I guess you could still input the data in a different format and then generate a lookup table on startup, but personally I didn't find having to type the class/specID in every entry to be a big enough annoyance to do anything about it. Also keep in mind that it supports cooldowns that are not tied to a class, like racials and items.

Without giving it a lot of though, I think the cleanest way to solve the dependency issue would be just to move all the spec detection code from GladiusEx to LCT. Then some of the logic that is currently in the cooldowns module could also be moved to the LCT, making the library easier to use and the cooldowns module code cleaner.

**Comment by vendethiel** [comment]
That's probably a good idea, but not something I'm willing to do in this PR, tbh.

**Comment by vendethiel** [comment]
Btw we already generate lookup tables by class/spec for the iterator at startup.
The layout will most probably stay the same.

**Comment by Rainrider** [comment]
My point of interest here was primarily why you introduce a circular dependency between GladiusEx and LCT.

You could take a look at the cooldowns module of oRA3 if you wish to improve LCT.

**Comment by slaren** [comment]
There is a big difference between detecting cooldowns of opponent players than of those in your own party. oRA3 is great but it is just not comparable.

**Comment by vendethiel** [comment]
> My point of interest here was primarily why you introduce a circular dependency between GladiusEx and LCT.

I know I do. The only other option is replacing the function (i.e. providing `LCT:SetSpecGetter` or whateverelse I could imagine), and it's not pretty either.

**Comment by vendethiel** [comment]
I tried to do that, but it just breaks everything, docs on the subject are non-existent, trying to follow what other repos do or reading LibStub docs is proving very annoying.

**Comment by vendethiel** [comment]
Okay, I actually managed to make it work.

Thanks for your help, re-reading all these comments helped me tremendously :-).

I need to fork LibCT and DRData and I'm done.

**Comment by vendethiel** [comment]
Pushed the cleanup.

File: GladiusEx.toc [hunk 8]
```diff
@@ -1,8 +1,8 @@
-﻿## Interface: 70200
+﻿## Interface: 70300
 ## Title: GladiusEx
 ## Notes: Arena unit frames
 ## SavedVariables: GladiusExDB
-## OptionalDeps: LibStub, LibFunctional-1.0, Ace3, DRData-1.0, LibSharedMedia-3.0, AceGUI-3.0-SharedMediaWidgets, AceGUI-3.0-Spell-EditBox, LibGroupInSpecT-1.1, LibCooldownTracker-1.0, LibDualSpec-1.0, LibRangeCheck-2.0, LibDispellable-1.0, Clique, Masque
+## OptionalDeps: LibStub, LibFunctional-1.0, Ace3, DRData-1.0, LibSharedMedia-3.0, AceGUI-3.0-SharedMediaWidgets, AceGUI-3.0-Spell-EditBox, LibGroupInSpecT-1.1, LibCooldownTracker-1.0, LibDualSpec-1.0, LibRangeCheck-2.0, LibDispellable-1.0, Clique, Masque, AceGUI-3.0-SharedMediaWidgets
```
Comment (slaren) [inline GladiusEx.toc]: AceGUI-3.0-SharedMediaWidgets was already on the list, it is duplicated now

**Comment by slaren** [comment]
I also noticed that you have copied the localization data to the repository. I don't mind if you want to stop using it, but before it used the [wowace localization tool](https://www.wowace.com/projects/gladiusex/localization), so that anyone could contribute translations.

**Comment by vendethiel** [comment]
It's less about downloading and more about "i started off the version from wowace"

File: GladiusEx.toc (hunk 8 above)
Comment (vendethiel) [reply GladiusEx.toc]: Interesting. Will remove.

**Comment by vendethiel** [comment]
Fixed the toc. How does the localization tool work?

**Comment by slaren** [comment]
You just have to add the strings that need to be localized to the localization tool, and then anyone can translate it to their language using the web app. The packager then automatically generates the translated locale tables.

**Comment by vendethiel** [comment]
Should be good to merge now!

**Comment by slaren** [comment]
Looks good to me, merge it whenever you want. Do you have a wowace/curseforge/curse/twitch/whatever it is called now username that I can add to the wowace project ?

**Comment by vendethiel** [comment]
yeah. Vendethiel.

**Comment by vendethiel** [comment]
🍰 🍰 🍰 🍰

**Comment by slaren** [comment]
Are you able to see the status of the packager at https://www.wowace.com/dashboard/builds ? The build failed with this error:
`Invalid pkgmeta file: [Line: 24, Col: 12] While scanning a simple key, could not find expected ':'.`

**Comment by vendethiel** [comment]
should be fixed

**Comment by slaren** [comment]
Failed again:
`Invalid pkgmeta file: [Line: 23, Col: 5] Duplicate key`

**Comment by vendethiel** [comment]
Ok. Should be fixed too.

Btw, I plan to declare "issue bankruptcy" and start anew, that is, close all issues open atm.

**Comment by vendethiel** [comment]
Ok, got the dashboard

**Comment by slaren** [comment]
Sure, do whatever you want with the issues. The build failed again:
`Unknown repository type, or unable to connect to repository. https://repos.wowace.com/wow/libdispellable-1-0`

**Comment by vendethiel** [comment]
Yup, I saw it go through. No issue with reading through CI logs :).

I want to take the time to say "thank you" for your supervision even though you really have no reason to care at all, during this month-and-so.

**Comment by vendethiel** [comment]
Whoo! Seems like it turned green

**Comment by vendethiel** [comment]
Aaaand it works!

**Comment by vendethiel** [comment]
@slaren How2Release? D:

**Comment by vendethiel** [comment]
@slaren About to release as 1.0.

**Comment by slaren** [comment]
Thank you for contributing to the project. It's the least I can do if someone is willing to contribute to my project rather than forking or starting their own.

You can have the packager make a release by creating a tag in the repository. There is some documentation about all this: https://www.curseforge.com/docs/packaging

**Comment by vendethiel** [comment]
Alright. Pushed a tag! Let's see.

**Comment by vendethiel** [comment]
Now I just need to figure out how to "push" the beta to mods.curse... Or do betas not get there?

**Comment by vendethiel** [comment]
Oh, is it because it's "under review"?

**Comment by slaren** [comment]
I am not sure why it is "under review", I guess some mod has to validate the project again.

**Comment by vendethiel** [comment]
It went through! 👍

**Pull request merged**